File: 1268.search-suggestions-system.py
#
# @lc app=leetcode id=1268 lang=python3
#
# [1268] Search Suggestions System
#

# @lc code=start
class Trie:
    def __init__(self):
        self.tree = {}
    def insert(self, word: str) -> None:
        chars = list(word)
        currDict = self.tree
        while chars:
            char = chars.pop(0)
            if not currDict.get(char):
                currDict[char] = {}
            currDict = currDict[char]
            
            if currDict.get("match"):
                currDict["match"].append(word)
            else:
                currDict["match"] = [word]
    def getTree(self):
        return self.tree

class Solution:
    def suggestedProducts(self, products: List[str], searchWord: str) -> List[List[str]]:
        trie = Trie()
        products = sorted(products) # sorted lexicographically
        for p in products:
            trie.insert(p)
        ans = []
        chars = list(searchWord)
        currDict = trie.getTree()
        while chars:
            char = chars.pop(0)
            if char not in currDict:
                break
            currDict = currDict[char]
            res = currDict["match"][:min(3, len(currDict["match"]))] if "match" in currDict else []
            ans.append(res)
        if len(ans)!=len(searchWord):
            for _ in range(len(searchWord)-len(ans)):
                ans.append([])
        return ans
        
# @lc code=end

File: 1268.search-suggestions-system.go
/*
 * @lc app=leetcode id=1268 lang=golang
 *
 * [1268] Search Suggestions System
 */

// @lc code=start
package main

import (
	"sort"
)

func insertString(ss []string, s string) []string {
	// Find the index where the string should be inserted
	i := sort.SearchStrings(ss, s)
	// Append a zero value to the slice to make room for the new string
	ss = append(ss, "")
	// Shift the elements to the right to create space for the new string
	copy(ss[i+1:], ss[i:])
	// Insert the new string at the found index
	ss[i] = s
	return ss
}

type Trie struct {
	_t       map[rune]*Trie // Change to pointer to Trie
	products []string
}

func newTrie() *Trie { // Return a pointer to Trie
	return &Trie{_t: make(map[rune]*Trie), products: []string{}}
}

func (this *Trie) Insert(val string) {
	currTrie := this // Use the pointer directly
	for _, letter := range val {
		if inner, ok := currTrie._t[letter]; ok == false {
			newTrie := newTrie()
			currTrie._t[letter] = newTrie
			currTrie = newTrie
		} else {
			currTrie = inner
		}
		// Append the product to the current node's products
		currTrie.products = insertString(currTrie.products, val)
	}
}

func suggestedProducts(products []string, searchWord string) [][]string {
	// 1. build a trie
	trie := newTrie()
	for _, product := range products {
		trie.Insert(product)
	}
	n := len(searchWord)
	res := make([][]string, n)
	for i := 0; i < n; i++ {
		res[i] = []string{}
	}
	currTrie := trie
	for i, letter := range searchWord {
		innerTrie, ok := currTrie._t[letter]
		if !ok {
			break
		}
		if len(innerTrie.products) >= 3 {
			res[i] = innerTrie.products[:3]
		} else {
			res[i] = innerTrie.products
		}
		currTrie = innerTrie // Move to the next node
	}
	return res
}

// @lc code=end
